common/src/com/thoughtworks/go/server/websocket/JsonMessage.java

```diff
@@ -0,0 +1,106 @@
+            JsonObject obj = json.getAsJsonObject();
+
+            Action action = Action.valueOf(obj.get("action").getAsString());
+            Object data = JsonMessage.deserialize(obj.get("data"));
+            return new Message(action, data);
+        }
+
+        @Override
+        public JsonElement serialize(Message src, Type typeOfSrc, JsonSerializationContext context) {
+            JsonObject obj = new JsonObject();
+            obj.addProperty("action", src.getAction().toString());
+            obj.add("data", JsonMessage.serialize(src.getData()));
+            return obj;
+        }
+    }
+
+    private static class ObjectConverter implements JsonSerializer<Object>, JsonDeserializer<Object> {
+        @Override
+        public Object deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
+            return JsonMessage.deserialize(json);
+        }
+
+        @Override
+        public JsonElement serialize(Object src, Type typeOfSrc, JsonSerializationContext context) {
+            return JsonMessage.serialize(src);
+        }
+    }
+
+    public static String encode(Message msg) {
+        return getGson().toJsonTree(msg).toString();
+    }
+
+    public static Message decode(String message) {
+        return getGson().fromJson(message, Message.class);
+    }
+
+    private static JsonElement serialize(Object src) {
+        if (src == null) {
+            return null;
+        }
+        JsonObject obj = new JsonObject();
+        obj.addProperty("type", src.getClass().getCanonicalName());
+        obj.add("data", getGson().toJsonTree(src));
+        return obj;
+    }
+
+    private static Object deserialize(JsonElement json) {
+        if (json == null) {
+            return null;
+        }
+        JsonObject object = json.getAsJsonObject();
+        String type = object.get("type").getAsString();
+        try {
+            Class k = Class.forName(type);
+            JsonElement data = object.get("data");
+            return getGson().fromJson(data.isJsonPrimitive() ? data.getAsJsonPrimitive() : data.getAsJsonObject(), k);
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static Gson getGson() {
+        GsonBuilder builder = new GsonBuilder()
+                .registerTypeAdapter(Builder.class, new ObjectConverter())
+                .registerTypeAdapter(JobPlan.class, new ObjectConverter())
+                .registerTypeAdapter(Material.class, new ObjectConverter())
+                .registerTypeAdapter(Message.class, new MessageConverter())
+                .registerTypeAdapter(FetchHandler.class, new ObjectConverter())
+                .registerTypeAdapter(MaterialInstance.class, new ObjectConverter())
+                .excludeFieldsWithModifiers(Modifier.TRANSIENT, Modifier.STATIC);
+        return builder.create();
```

Comment:
Gson can be an expensive object to create — It is threadsafe and is safe to cache at a class level (aka singleton)